sonar-batch/src/main/java/org/sonar/batch/bootstrap/TempFolderProvider.java

```diff
@@ -19,33 +19,109 @@
+  static final String TMP_NAME_PREFIX = ".sonartmp_";
 
-  private TempFolder tempFolder;
+  private DefaultTempFolder tempFolder;
 
   public TempFolder provide(BootstrapProperties bootstrapProps) {
     if (tempFolder == null) {
-      String workingDirPath = StringUtils.defaultIfBlank(bootstrapProps.property(CoreProperties.WORKING_DIRECTORY), CoreProperties.WORKING_DIRECTORY_DEFAULT_VALUE);
-      File workingDir = new File(workingDirPath).getAbsoluteFile();
-      File tempDir = new File(workingDir, ".sonartmp");
+
+      String workingPathName = StringUtils.defaultIfBlank(bootstrapProps.property(CoreProperties.GLOBAL_WORKING_DIRECTORY), CoreProperties.GLOBAL_WORKING_DIRECTORY_DEFAULT_VALUE);
+      Path workingPath = Paths.get(workingPathName).normalize();
+
+      if (!workingPath.isAbsolute()) {
+        Path home = findHome(bootstrapProps);
+        workingPath = home.resolve(workingPath).normalize();
+      }
+
+      try {
+        cleanTempFolders(workingPath);
+      } catch (IOException e) {
+        LOG.warn("failed to clean global working directory: " + e.getMessage());
+      }
+
+      Path tempDir = workingPath.resolve(TMP_NAME_PREFIX + System.currentTimeMillis());
       try {
-        FileUtils.forceMkdir(tempDir);
+        Files.createDirectories(tempDir);
       } catch (IOException e) {
         throw new IllegalStateException("Unable to create root temp directory " + tempDir, e);
       }
-      tempFolder = new DefaultTempFolder(tempDir);
+      tempFolder = new DefaultTempFolder(tempDir.toFile(), true);
+      this.instance = tempFolder;
     }
     return tempFolder;
   }
 
+  private static Path findHome(BootstrapProperties props) {
+    String home = props.property("sonar.userHome");
+    if (home != null) {
+      return Paths.get(home);
+    }
+
+    home = System.getenv("SONAR_USER_HOME");
+
+    if (home != null) {
+      return Paths.get(home);
+    }
+
+    home = System.getProperty("user.home");
+    return Paths.get(home, ".sonar");
+  }
+
+  private static void cleanTempFolders(Path path) throws IOException {
+    if (Files.exists(path)) {
+      try (DirectoryStream<Path> stream = Files.newDirectoryStream(path, new CleanFilter())) {
+        for (Path p : stream) {
+          FileUtils.deleteQuietly(p.toFile());
+        }
+      }
+    }
+  }
+
+  private static class CleanFilter implements DirectoryStream.Filter<Path> {
+    @Override
+    public boolean accept(Path e) throws IOException {
+      if (!Files.isDirectory(e)) {
+        return false;
+      }
+
+      if (!e.getFileName().toString().startsWith(TMP_NAME_PREFIX)) {
+        return false;
+      }
+
+      long threshold = System.currentTimeMillis() - CLEAN_MAX_AGE;
+
+      // we could also check the timestamp in the name, instead
+      BasicFileAttributes attrs;
+
+      try {
+        attrs = Files.readAttributes(e, BasicFileAttributes.class);
+      } catch (IOException ioe) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)